app/src/main/java/org/transdroid/search/AbstractHtmlAdapter.java

```diff
@@ -0,0 +1,177 @@
+
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.entity.UrlEncodedFormEntity;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.impl.client.BasicCookieStore;
+import org.apache.http.impl.client.DefaultHttpClient;
+import org.apache.http.impl.cookie.BasicClientCookie;
+import org.apache.http.message.BasicNameValuePair;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+import org.transdroid.search.gui.SettingsHelper;
+import org.transdroid.util.HttpHelper;
+
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A base class for HTML scraping adapters. It handles most of the boilerplate code.
+ */
+public abstract class AbstractHtmlAdapter implements ISearchAdapter {
+
+    @Override
+    public List<SearchResult> search(SharedPreferences prefs, String query, SortOrder order, int maxResults) throws Exception {
+        // todo: All adapters that use this method call with ignoreSslIssues=false. Perhaps it needs
+        // If an adapter that need ignoreSslIssues=true comes around, we can extract an abstract
+        // method to rpovide it.
+        final DefaultHttpClient client = HttpHelper.buildDefaultSearchHttpClient(false);
+        authenticateHttpClient(client, prefs);
+
+        final String searchUrl = getSearchUrl(prefs, query, order, maxResults);
+        final HttpGet get = new HttpGet(searchUrl);
+        final HttpResponse response = client.execute(get);
+        final HttpEntity entity = response.getEntity();
+        final Document document = Jsoup.parse(entity.getContent(), null, "");
+        entity.consumeContent();
+
+        final Elements torrentElements = selectTorrentElements(document);
+        final ArrayList<SearchResult> searchResults = new ArrayList<>();
+        for (Element torrentElement : torrentElements) {
+            searchResults.add(buildSearchResult(torrentElement));
+            if (searchResults.size() == maxResults) {
+                break;
+            }
+        }
+        return searchResults;
+    }
+
+    @Override
+    public InputStream getTorrentFile(SharedPreferences prefs, String url) throws Exception {
+        final DefaultHttpClient httpClient = new DefaultHttpClient();
+        if (isAuthenticationRequiredForTorrentLink()) {
+            authenticateHttpClient(httpClient, prefs);
+        }
+        final HttpGet request = new HttpGet(url);
+
+        final HttpResponse response = httpClient.execute(request);
+        return response.getEntity().getContent();
+    }
+
+    /**
+     *
+     * @return true is this site generated authenticated link. false if the links require manual
+     * authentication.
+     */
+    @SuppressWarnings("WeakerAccess")
```

Comment:
It's just so dev tools (like IntelliJ) doesn't complain about unused methods etc'
Some of the methods are still unused because the abstract class implementation is sufficient.

Reply:
Hmm with `WeakerAccess`? Whatever, we will see.